### PR TITLE
Implements quitgame function

### PR DESCRIPTION
**Pull request by dethrace-labs**
- `QuitGame` was only partially implemented
- Added extra S3 functions required for shutting down
- Moved S3 device opening functions to something more closely matching OG code

### CODE REVIEW

File: src/DETHRACE/common/main.c [hunk 1]
```diff
@@ -27,7 +29,21 @@ void QuitGame() {
...
+    ShutdownNetIfRequired();
+    if (gSound_available) {
+        DRS3ShutDown();
+    }
+    if (gBr_initialized) {
+        ClearEntireScreen();
+    }
+    PDRevertPalette();
+    StopMusic();
+    PDShutdownSystem();
```
Comment (madebr) [inline src/DETHRACE/common/main.c:46]: I can also see `CloseDiagnostics()` and `exit(0)`, even though `PDShutdownSystem()` on Windows will already exit the system.
But I'm not sure
```suggestion
    PDShutdownSystem();
    CloseDiagnostics();
    exit(0);
```

File: src/DETHRACE/pc-dos/dossys.c [hunk 2]
```diff
@@ -308,7 +308,8 @@ void PDInitialiseSystem() {
 // IDA: void __cdecl PDShutdownSystem()
 void PDShutdownSystem() {
     LOG_TRACE("()");
-    NOT_IMPLEMENTED();
+
+    exit(0);
```
Comment (madebr) [inline src/DETHRACE/pc-dos/dossys.c:312]: Not needed now, but perhaps we need to add a harness hook here?
We can then destroy the OpenGL and SDL contexts.

Comment (dethrace-labs) [reply src/DETHRACE/pc-dos/dossys.c:315]: good point, yes, added it 👍

File: src/DETHRACE/common/main.c (hunk 1 above)
Comment (dethrace-labs) [reply src/DETHRACE/common/main.c:46]: good spotting! For some reason, IDA was confused by this function and the disassembler was missing the last bit of it. Good reminder to not be lazy and double check the assembly too :)

Comment (madebr) [reply src/DETHRACE/common/main.c:46]: In that regard, ghidra is moody too.
It does not always show what happens after `FatalError`, but the decompiler showed everything here.
I suppose you need to always be careful when marking a function as "does not return".

File: src/harness/io_platforms/sdl_gl.c [hunk 3]
```diff
@@ -203,6 +203,12 @@ tRenderer* Window_Create(char* title, int width, int height, int pRender_width,
     return &gl_renderer;
 }
 
+void Window_Close() {
+    SDL_GL_DeleteContext(context);
+    SDL_DestroyWindow(window);
+    window = NULL;
```
Comment (madebr) [inline src/harness/io_platforms/sdl_gl.c:209]: You can also do `SDL_Quit()`.

File: src/harness/harness.c [hunk 4]
```diff
@@ -317,6 +317,10 @@ void Harness_Hook_GraphicsInit(int render_width, int render_height) {
     renderer = Window_Create("Dethrace", window_width, window_height, render_width, render_height);
 }
 
+void Harness_Hook_PDShutdownSystem() {
+    Window_Close();
```
Comment (madebr) [inline src/harness/harness.c:321]: The platform also must shut down, in this case SDL.

File: docs/PORTING.md [hunk 5]
```diff
@@ -38,6 +38,7 @@ To add a new `IOPlatform`:
 - `Window_Create`
 - `Window_PollEvents`
 - `Window_Swap`
+- `IOPlatform_Shutdown`
 - `Input_GetKeyMap`
 - `Input_IsKeyDown`
```
Comment (madebr) [inline docs/PORTING.md:43]: Should all these functions have a `IOPlatform_` prefix? Right now it's a mishmash. (something to think about)

**Review by madebr — APPROVED** [review]

**Pull request merged**